cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,304 @@
+				fc.ensureReconciledFirewall(ctx)
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
+	defer cancel()
+
+	wait.Until(func() {
+		currentFirewall, err := fc.fwManager.Get(ctx)
+		if err != nil {
+			klog.Errorf("failed to get worker firewall: %s", err)
+		}
+
+		if currentFirewall != nil {
+			fc.fwCache.mu.RLock()
+			isEqual := cmp.Equal(currentFirewall, fc.fwCache.firewall)
+			fc.fwCache.mu.RUnlock()
+
+			if isEqual {
+				return
+			}
+		}
+		err = fc.ensureReconciledFirewall(ctx)
+		if err != nil {
+			klog.Errorf("failed to reconcile worker firewall: %s", err)
+		}
+
+	}, firewallReconcileRequency, stopCh)
+}
+
+// Get returns the current public access firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	if fm.fwCache.firewall != nil {
+		fm.fwCache.mu.RLock()
+		fwID := fm.fwCache.firewall.ID
+		fm.fwCache.mu.RUnlock()
+
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fwID)
+		if err != nil && (resp == nil || resp.StatusCode != http.StatusNotFound) {
+			return nil, fmt.Errorf("failed to retrieve firewall by ID: %s", err)
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			klog.Warningf("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return cmp.Equal(fw.Name, fm.workerFirewallName)
+	}
+	klog.Info("filtering firewall list for the firewall that has the expected firewall name")
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %s", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the public access firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	fm.fwCache.mu.RLock()
+	targetFirewall := fm.fwCache.firewall
+	fm.fwCache.mu.RUnlock()
+
+	fwSvc := fm.client.Firewalls
+
+	fm.fwCache.mu.RLock()
+	isEqual := cmp.Equal(targetFirewall.InboundRules, svcInboundRules)
+	fm.fwCache.mu.RUnlock()
+
+	if isEqual {
+		return nil
+	}
+
+	// Check if the target firewall ID exists but in the case that CCM first starts up and the
+	// firewall ID does not exist yet, we have to retrieve one from the API.
+	var fwID string
+	if targetFirewall == nil {
+		currentFirewall, err := fm.Get(ctx)
+		if err != nil {
+			// Log error then create firewall and update cache, since firewall does not exist.
```

Comment:
`fm.Get` returns an error if `filterFirewallList` returns an error, which doesn't mean the firewall is not found. I think the case where we actually want to create is `if err == nil && currentFirewall == nil`. Am I missing anything?

Reply:
The assumption was that if an error gets returned, it will not include a firewall. Based on filterFirewallList
```go
func filterFirewallList(ctx context.Context, client *godo.Client, matchExpectedFirewallName func(godo.Firewall) bool) (*godo.Firewall, error) {
	opt := &godo.ListOptions{Page: 1, PerPage: apiResultsPerPage}
	for {
		firewalls, resp, err := client.Firewalls.List(ctx, opt)
		if err != nil {
			return nil, err
		}

		for _, fw := range firewalls {
			if matchExpectedFirewallName(fw) {
				return &fw, nil
			}
		}

		// if we are at the last page, break out the for loop
		if resp.Links == nil || resp.Links.IsLastPage() {
			break
		}

		page, err := resp.Links.CurrentPage()
		if err != nil {
			return nil, err
		}

		opt.Page = page + 1
	}

	return nil, nil
}
```

and its calls out to the firewall code:
```go
// List Firewalls.
func (fw *FirewallsServiceOp) List(ctx context.Context, opt *ListOptions) ([]Firewall, *Response, error) {
	path, err := addOptions(firewallsBasePath, opt)
	if err != nil {
		return nil, nil, err
	}

	return fw.listHelper(ctx, path)
}
```

and

```go
func (fw *FirewallsServiceOp) listHelper(ctx context.Context, path string) ([]Firewall, *Response, error) {
	req, err := fw.client.NewRequest(ctx, http.MethodGet, path, nil)
	if err != nil {
		return nil, nil, err
	}

	root := new(firewallsRoot)
	resp, err := fw.client.Do(ctx, req, root)
	if err != nil {
		return nil, resp, err
	}
	if l := root.Links; l != nil {
		resp.Links = l
	}
	if m := root.Meta; m != nil {
		resp.Meta = m
	}

	return root.Firewalls, resp, err
}
```

It looks like when there is a _possible_ opportunity to return both, this is returned `return nil, nil, err` where `firewall` and `response` is nil and `err` is not nil. Therefore, it _looks_ like, the firewall service List will not return both a `firewall` and an `err` (unless I am misreading one of these functions).


If I understand the question correctly:
> I think the case where we actually want to create is if err == nil && currentFirewall == nil. Am I missing anything?

It is correct that the current code's assumption is that it is not found if an error comes back. Should I add a check as a form of defensive programming in the case that the API changes to do something we didn't originally expect? It removes the assumption with an explicit check, which I think makes sense.

Reply:
You're right that godo won't return both a firewall and an error. However, an error return doesn't mean the firewall doesn't exist, and the absence of an error doesn't mean the firewall does exist. So, there are two scenarios I'm concerned about:

1. Suppose the firewall does exist, but there's a transient DO API issue or a network blip. `filterFirewallList` would be unable to find the firewall and return an error. As the code is right now, that would result in CCM attempting to create a *second* firewall. The create will fail since the name isn't unique, so it's not the end of the world, but it would be easier to debug if we returned the error from `Get` rather than proceeding to try and create.

2. If the firewall really doesn't exist, `filterFirewallList` will return `nil, nil`. In this case, we won't try to create a new firewall, `fwID` won't be set, and the `Update` below will fail, and we'll create the firewall then.

Reply:
Ok, this makes more sense. Thank you for the clarification!